be/src/exprs/time_functions.cpp

```diff
@@ -2378,6 +2378,151 @@ StatusOr<ColumnPtr> TimeFunctions::make_date(FunctionContext* context, const Col
+}
+
+StatusOr<ColumnPtr> TimeFunctions::date_diff_hour(FunctionContext* context, const Columns& columns) {
+    ColumnViewer<TYPE_DATETIME> lv_column(columns[0]);
+    ColumnViewer<TYPE_DATETIME> rv_column(columns[1]);
+    ColumnViewer<TYPE_VARCHAR> type_column(columns[2]);
+    auto size = columns[2]->size();
+    ColumnBuilder<TYPE_BIGINT> result(size);
+    for (int row = 0; row < size; ++row) {
+        TimestampValue l = (TimestampValue)lv_column.value(row);
+        TimestampValue r = (TimestampValue)rv_column.value(row);
+        auto type_str = type_column.value(row).to_string();
+        result.append(l.diff_microsecond(r) / USECS_PER_HOUR);
+    }
+    return result.build(ColumnHelper::is_all_const(columns));
+}
+
+StatusOr<ColumnPtr> TimeFunctions::date_diff_minute(FunctionContext* context, const Columns& columns) {
+    ColumnViewer<TYPE_DATETIME> lv_column(columns[0]);
+    ColumnViewer<TYPE_DATETIME> rv_column(columns[1]);
+    ColumnViewer<TYPE_VARCHAR> type_column(columns[2]);
+    auto size = columns[2]->size();
+    ColumnBuilder<TYPE_BIGINT> result(size);
+    for (int row = 0; row < size; ++row) {
+        TimestampValue l = (TimestampValue)lv_column.value(row);
+        TimestampValue r = (TimestampValue)rv_column.value(row);
+        auto type_str = type_column.value(row).to_string();
+        result.append(l.diff_microsecond(r) / USECS_PER_MINUTE);
+    }
+    return result.build(ColumnHelper::is_all_const(columns));
+}
+
+StatusOr<ColumnPtr> TimeFunctions::date_diff_second(FunctionContext* context, const Columns& columns) {
+    ColumnViewer<TYPE_DATETIME> lv_column(columns[0]);
+    ColumnViewer<TYPE_DATETIME> rv_column(columns[1]);
+    ColumnViewer<TYPE_VARCHAR> type_column(columns[2]);
+    auto size = columns[2]->size();
+    ColumnBuilder<TYPE_BIGINT> result(size);
+    for (int row = 0; row < size; ++row) {
+        TimestampValue l = (TimestampValue)lv_column.value(row);
+        TimestampValue r = (TimestampValue)rv_column.value(row);
+        auto type_str = type_column.value(row).to_string();
+        result.append(l.diff_microsecond(r) / USECS_PER_SEC);
+    }
+    return result.build(ColumnHelper::is_all_const(columns));
+}
+
+StatusOr<ColumnPtr> TimeFunctions::date_diff_millisecond(FunctionContext* context, const Columns& columns) {
+    ColumnViewer<TYPE_DATETIME> lv_column(columns[0]);
+    ColumnViewer<TYPE_DATETIME> rv_column(columns[1]);
+    ColumnViewer<TYPE_VARCHAR> type_column(columns[2]);
+    auto size = columns[2]->size();
+    ColumnBuilder<TYPE_BIGINT> result(size);
+    for (int row = 0; row < size; ++row) {
+        TimestampValue l = (TimestampValue)lv_column.value(row);
+        TimestampValue r = (TimestampValue)rv_column.value(row);
+        auto type_str = type_column.value(row).to_string();
+        result.append(l.diff_microsecond(r) / USECS_PER_MILLIS);
+    }
+    return result.build(ColumnHelper::is_all_const(columns));
+}
```

Comment:
too many redundant code in date_diff_xx function, suggest use template or macro to simplify it

Reply:
ok，i would use one function to handle these conditions